osu.Game/Screens/Play/Player.cs

```diff
@@ -289,6 +289,12 @@ private bool onFail()
             if (Beatmap.Value.Mods.Value.OfType<IApplicableFailOverride>().Any(m => !m.AllowFail))
                 return false;
 
+            if (Beatmap.Value.Mods.Value.OfType<IApplicableFailOverride>().Any(m => m.RestartOnFail))
+            {
+                Restart();
+                return false;
```

Comment:
This should `return true` otherwise `onFail()` is going to be called multiple times.